spark/src/test/scala/org/apache/spark/sql/delta/schema/SchemaUtilsSuite.scala

```diff
@@ -2569,6 +2573,244 @@ class SchemaUtilsSuite extends QueryTest
+        AllTypeWidening,
+        TypeEvolution(uniformIcebergCompatibleOnly = false),
+        TypeEvolution(uniformIcebergCompatibleOnly = true),
+        AllTypeWideningToCommonWiderType,
+        TypeEvolutionToCommonWiderType(uniformIcebergCompatibleOnly = false),
+        TypeEvolutionToCommonWiderType(uniformIcebergCompatibleOnly = true))) {
+      // byte, short, int are all stored as INT64 in parquet, [[mergeSchemas]] always allows
+      // widening between them. This was already the case before typeWideningMode was introduced.
+      val merged1 = mergeSchemas(narrow, wide, typeWideningMode = typeWideningMode)
+      assert(merged1 === wide)
+      val merged2 = mergeSchemas(wide, narrow, typeWideningMode = typeWideningMode)
+      assert(merged2 === wide)
+    }
+  }
+
+  // These type changes will only be available once Delta uses Spark 4.0.
+  for ((fromType, toType) <- Seq(
+    IntegerType -> LongType,
+    new StructType().add("x", IntegerType) -> new StructType().add("x", LongType),
+    MapType(IntegerType, IntegerType) -> MapType(LongType, LongType),
+    ArrayType(IntegerType) -> ArrayType(LongType)
+  ))
+  testSparkMasterOnly(s"typeWideningMode ${fromType.sql} -> ${toType.sql}") {
+    import TypeWideningMode._
+
+    val narrow = new StructType().add("a", fromType)
+    val wide = new StructType().add("a", toType)
+
+    for (typeWideningMode <- Seq(
+        NoTypeWidening,
+        AllTypeWidening,
+        TypeEvolution(uniformIcebergCompatibleOnly = false),
+        TypeEvolution(uniformIcebergCompatibleOnly = true))) {
+      // Narrowing is not allowed.
+      expectAnalysisErrorClass("DELTA_MERGE_INCOMPATIBLE_DATATYPE",
+        Map("currentDataType" -> "LongType", "updateDataType" -> "IntegerType")) {
+        mergeSchemas(wide, narrow, typeWideningMode = typeWideningMode)
+      }
+    }
+
+    for (typeWideningMode <- Seq(
+        AllTypeWideningToCommonWiderType,
+        TypeEvolutionToCommonWiderType(uniformIcebergCompatibleOnly = false),
+        TypeEvolutionToCommonWiderType(uniformIcebergCompatibleOnly = true))) {
+      // Bidirectional doesn't enforce an order on the inputs, widening from second schema to first
+      // is allowed.
+      val merged = mergeSchemas(wide, narrow, typeWideningMode = typeWideningMode)
+      assert(merged === wide)
+    }
+
+    for (typeWideningMode <- Seq(
+        AllTypeWidening,
+        TypeEvolution(uniformIcebergCompatibleOnly = false),
+        TypeEvolution(uniformIcebergCompatibleOnly = true),
+        AllTypeWideningToCommonWiderType,
+        TypeEvolutionToCommonWiderType(uniformIcebergCompatibleOnly = false),
+        TypeEvolutionToCommonWiderType(uniformIcebergCompatibleOnly = true))) {
+      // Widening is allowed, unless mode is NoTypeWidening.
+      val merged = mergeSchemas(narrow, wide, typeWideningMode = typeWideningMode)
+      assert(merged === wide)
+    }
+    expectAnalysisErrorClass("DELTA_MERGE_INCOMPATIBLE_DATATYPE",
```

Comment:
is this not duplicate as it is also handled via line 2625 and checked in 2630?

Reply:
No, previous one covers `narrow` -> `wide`, this one covers `wide` -> `narrow`